### DISCUSSION TITLE
ip白名单策略原理

### DISCUSSION THREAD

**Discussion by yydance** [apache/apisix#8071] [Q&A]
问题描述：
apisix位于LB或者nginx后，使用IP白名单限制访问路由，结果未生效，大概看了下插件的源码，似乎是匹配的remote_addr，不知道是否正确？因为我不是研发，lua以前学过一点语法很久没看了。如果取自remote_addr，似乎不是太合适，是基于什么考虑不取x-real-ip，尽管x-real-ip可能不存在，但是我们可以默认赋值remote_addr。

**Comment by tzssangglass** [comment]
The default is to use `X-Real-IP` in headers as `remote_addr`, ref: https://github.com/apache/apisix/blob/c201d72e398a08d3e36695f89576410afe541365/conf/config-default.yaml#L200

verify:

route:

```shell
curl http://127.0.0.1:9180/apisix/admin/routes/1 -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -d '
{
    "uri": "/hello",
    "upstream": {
        "type": "roundrobin",
        "nodes": {
            "127.0.0.1:1980": 1
        }
    },
    "plugins": {
        "ip-restriction": {
            "whitelist": [
                "127.0.0.1"
            ]
        }
    }
}'
```

send request from `127.0.0.1`

```shell
curl http://127.0.0.1:9080/hello                                                                       '
1234567890#
```

send request from `127.0.0.1` and with wrong `X-Real-IP`, ip-restriction works well

```shell
curl http://127.0.0.1:9080/hello -H "X-Real-IP: 192.168.0.1"
{"message":"Your IP address is not allowed"}
```

**Reply by yydance** [reply]
白名单策略：
```
"message": "Your IP address is not allowed",
      "whitelist": [
        "10.0.0.0/8",
        "172.16.0.0/16",
        "192.168.0.0/16",
        "127.0.0.1",
        "101.x.x.x",
        "119.x.x.x",
        "82.x.x.x",
        "106.x.x.x",
        "106.x.x.x",
        "101.x.x.x",
        "124.x.x.x",
        "82.x.x.x",
        "164.x.x.x",
        "114.x.x.x",
        "43.134.53.175",
        "148.x.x.x",
        "124.x.x.x"
      ]
```
我的日志格式，前几个字段：
```
[$time_local] $http_X_Real_IP $remote_addr
```
请求的日志结果：
```
[12/Oct/2022:11:01:06 +0800] 43.129.176.124 10.8.112.69 10.1.91.111:61183 GET 200 374 - HTTP/1.1 "http://console.camin.vip" "/" "-" "curl/7.29.0" "-" 0.009 0.009 200 "http://10.1.91.111:61183" tcdn console.camin.vip
```

x-real-ip是43.129.176.124 ，并没有block

**Reply by tzssangglass** [reply]
which APISIX get the `$remote_addr` maybe `10.8.112.69`(maybe your Nginx's or LB's IP), and you have `"10.0.0.0/8"` in whitelist

**Reply by yydance** [reply]
您这条回复是否表明白名单match的是$remote_addr

**Reply by yydance** [reply]
你把示例中请求地址http://127.0.0.1:9080/hello 换成http://{eth0_ip}:9080/hello -H 'X-Real-IP: 127.0.0.1' 就过不去了

**Reply by tzssangglass** [reply]
then we need use `real-ip` plugin, for example, my `eth0_ip` is `10.211.55.4`

```
curl http://127.0.0.1:9180/apisix/admin/routes/1 -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -d '
{
    "uri": "/hello",
    "upstream": {
        "type": "roundrobin",
        "nodes": {
            "127.0.0.1:1980": 1
        }
    },
    "plugins": {
        "ip-restriction": {
            "whitelist": [
                "127.0.0.1"
            ]
        },
        "real-ip": {
            "source": "http_x_real_ip",
            "trusted_addresses": ["10.211.55.4"]
        }
    }
}'
```

and test

```
curl http://10.211.55.4:9080/hello -H "X-Real-IP: 127.0.0.1"
```

this work for me

**Reply by tzssangglass** [reply]
if you don't want use `real-ip` plugin, you can add

```yaml
nginx_config: 
  http:
    real_ip_from:
      - ${eth0_ip}
```

**Comment by yydance** [comment]
另一个示例
1、白名单
```
{
  "disable": false,
  "whitelist": [
    "127.0.0.1"
  ]
}
```
2、请求
```
curl -H 'host:prom.camin.vip' -H 'X-Real-IP: 127.0.0.1'  http://10.0.2.137:61160
{"message":"Your IP address is not allowed"}
```
3、打印的日志
```
[12/Oct/2022:12:12:36 +0800] 127.0.0.1 10.0.13.111 - GET 403 56 - HTTP/1.1 "http://prom.camin.vip" "/" "-" "curl/7.29.0" "-" 0.000 - - "http://prom.camin.vip" - -
```

**Reply by yydance** [reply]
当我白名单加了IP 10.0.13.111之后就可以访问了

**Reply by tzssangglass** [reply]
which APISIX gets the `$remote_addr` maybe 10.0.13.111(maybe your Nginx's or LB's IP).

**Reply by yydance** [reply]
10.0.13.111 是发起curl请求的客户端IP

**Comment by yydance** [comment]
@tzssangglass 还请百忙之中抽点时间解疑，谢谢。

**Reply by tzssangglass** [reply]
read more: https://github.com/apache/apisix/discussions/4793

**Comment by yydance** [comment]
我知道如何获取X-Real-IP。这里综上的例子，看起来IP白名单whitelist策略match的是$remote_addr，而不是取X-Real-IP 的值 @tzssangglass

**Reply by tokers** [reply]
Which Apache APISIX version are you using?

**Reply by yydance** [reply]
apisix: 2.15
apisix-dashboard: 2.13

**Reply by tzssangglass** [reply]
> 我知道如何获取X-Real-IP。这里综上的例子，看起来IP白名单whitelist策略match的是$remote_addr，而不是取X-Real-IP 的值

But I gave example proves that the remote_addr gets the lower X-Real-IP?

**Reply by yydance** [reply]
我觉得remote_addr不应该使用x-real-ip赋值，应该是x-real-ip不存在的情况下使用remote_addr赋值，whitelist match的IP应该取自x-real-ip

**Comment by tzssangglass** [comment]
You should be concerned about getting your front-end Nginx or LB to carry real client IP to APISIX

**Comment by yydance** [comment]
前边的日志格式和日志内容表明 real client ip 是正确的，源代码init.lua里也没看到有关real ip：
```
function _M.restrict(conf, ctx)
    local block = false
    local remote_addr = ctx.var.remote_addr

    if conf.blacklist then
        local matcher = lrucache(conf.blacklist, nil,
                                 core.ip.create_ip_matcher, conf.blacklist)
        if matcher then
            block = matcher:match(remote_addr)
        end
    end

    if conf.whitelist then
        local matcher = lrucache(conf.whitelist, nil,
                                 core.ip.create_ip_matcher, conf.whitelist)
        if matcher then
            block = not matcher:match(remote_addr)
        end
    end

    if block then
        return 403, { message = conf.message }
    end
end
```

**Comment by yydance** [comment]
综合上述的信息，看起来IP白名单插件使用的是$remote_addr match whitelist，然而给的示例似乎是将x-real-ip赋值给了$remote_addr，如果属实，这个逻辑是不对的，白名单IP应该使用x-real-ip match whitelist，当x-real-ip值为空时，默认赋值$remote_addr，$remote_addr获取的IP应该始终是request client ip，不应该被更改。

**Reply by tzssangglass** [reply]
Why does `ip-restriction` use `$remote_addr` instead of `X-Real-IP` header to restrict?

1. The `$remote_addr` variable refers to the real client IP, but the real client IP may be in the `X-Real-IP` header, or it may be in the `X-Forward-For` header, or it may be in some other special header (some cloud LBs do this)

> 然而给的示例似乎是将x-real-ip赋值给了$remote_addr，如果属实，这个逻辑是不对的

In your scenario, APISIX has proxies in front of it, and it seems that APISIX does not assign the real client IP to the variable `$remote_addr`. You can change the default configuration here: https://github.com/apache/apisix/blob/6ae923a12fc1b7f8414b66efc053b19c7b5ddd08/conf/config-default.yaml#L200-L204, or rescue the [real-ip](https://github.com/apache/apisix/blob/master/docs/zh/latest/plugins/real-ip.md) plugin to ensure that `$remote_addr` points to the real client IP.

**Reply by yydance** [reply]
在案例 https://github.com/apache/apisix/discussions/8071#discussioncomment-3856291 里没有经过任何代理，直接访问的apisix，结果与您给的示例是相反的，稍后我也将启用real-ip插件进行测试。

**Reply by tzssangglass** [reply]
> 在案例 https://github.com/apache/apisix/discussions/8071#discussioncomment-3856291 里没有经过任何代理，直接访问的apisix，结果与您给的示例是相反的，稍后我也将启用real-ip插件进行测试。

I can't reproduce it, I think you need to explore it in your environment.

**Reply by yydance** [reply]
real ip插件也是不行。
环境：
docker-compose部署，来自官方git仓库，所有操作均在dashboard，real_ip_header: X-Real-IP，使用curl 发起请求，从日志看，whitelist 匹配的是$remote_addr的值
我觉得，让其他人试下，看是否有问题

**Reply by tzssangglass** [reply]
BTW, I didn't deploy with docker-compose. I tested it directly on the host computer.

**Reply by tzssangglass** [reply]
Deployed with docker-compose, APISIX runs inside docker and traffic requesting APISIX on the host will come in from the docker's NIC.

**Comment by yydance** [comment]
刚从官方git仓库拉了个最新版，dashboard 2.13, apisix 2.99 ，什么都没更改，请求没经过任何代理，结果仍然是不行，whitelist中匹配到remote_addr才可行，不清楚您给的示例是什么配置

**Reply by tzssangglass** [reply]
> 不清楚您给的示例是什么配置

My example already gives the configuration.

**Comment by tzssangglass** [comment]
I think you should focus your question on understanding how APISIX gets the real client IP. Refer to the link I gave above.

Again, it makes sense for the ip-restriction plugin to use the $remote_addr variable to get the real client IP.

**Reply by yydance** [reply]
宿主机请求访问，经过多种测试，只有请求http://127.0.0.1:9080/的形式，remote_addr会被替换成X-Real-IP的值(如果存在)

**Reply by tzssangglass** [reply]
> 宿主机请求访问，经过多种测试，只有请求[http://127.0.0.1:9080/的形式，remote_addr会被替换成X-Real-IP的值(如果存在)](http://127.0.0.1:9080/%E7%9A%84%E5%BD%A2%E5%BC%8F%EF%BC%8Cremote_addr%E4%BC%9A%E8%A2%AB%E6%9B%BF%E6%8D%A2%E6%88%90X-Real-IP%E7%9A%84%E5%80%BC(%E5%A6%82%E6%9E%9C%E5%AD%98%E5%9C%A8))

I think you still don't understand the logic of how the real-ip module works in Nginx

**Reply by yydance** [reply]
你是对的，整清楚了，非常感谢 @tzssangglass
